module/os/linux/zfs/zpl_file.c

```diff
@@ -998,6 +998,117 @@ zpl_ioctl_setxattr(struct file *filp, void __user *arg)
 	return (err);
 }
 
+/*
+ * Expose Additional File Level Attributes of ZFS.
+ */
+static int
+zpl_ioctl_getdosflags(struct file *filp, void __user *arg)
+{
+	struct inode *ip = file_inode(filp);
+	uint64_t dosflags = ITOZ(ip)->z_pflags;
+	dosflags &= ZFS_DOS_FL_USER_VISIBLE;
+	int err = copy_to_user(arg, &dosflags, sizeof (dosflags));
+
+	return (err);
+}
+
+static int
+__zpl_ioctl_setdosflags(struct inode *ip, uint64_t ioctl_flags, xvattr_t *xva)
+{
+	uint64_t zfs_flags = ITOZ(ip)->z_pflags;
+	xoptattr_t *xoap;
+
+	if (ioctl_flags & (~ZFS_DOS_FL_USER_VISIBLE))
+		return (-EOPNOTSUPP);
+
+	if ((fchange(ioctl_flags, zfs_flags, ZFS_IMMUTABLE, ZFS_IMMUTABLE) ||
+	    fchange(ioctl_flags, zfs_flags, ZFS_APPENDONLY, ZFS_APPENDONLY)) &&
+	    !capable(CAP_LINUX_IMMUTABLE))
+		return (-EPERM);
+
+	if (!zpl_inode_owner_or_capable(kcred->user_ns, ip))
+		return (-EACCES);
+
+	xva_init(xva);
+	xoap = xva_getxoptattr(xva);
+
+	XVA_SET_REQ(xva, XAT_IMMUTABLE);
+	if (ioctl_flags & ZFS_IMMUTABLE)
+		xoap->xoa_immutable = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_APPENDONLY);
+	if (ioctl_flags & ZFS_APPENDONLY)
+		xoap->xoa_appendonly = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_NODUMP);
+	if (ioctl_flags & ZFS_NODUMP)
+		xoap->xoa_nodump = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_READONLY);
+	if ((ioctl_flags & ZFS_READONLY))
+		xoap->xoa_readonly = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_HIDDEN);
+	if ((ioctl_flags & ZFS_HIDDEN))
+		xoap->xoa_hidden = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_SYSTEM);
+	if ((ioctl_flags & ZFS_SYSTEM))
+		xoap->xoa_system = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_ARCHIVE);
+	if ((ioctl_flags & ZFS_ARCHIVE))
+		xoap->xoa_archive = B_TRUE;
+
+	XVA_SET_REQ(xva, XAT_NOUNLINK);
+	if ((ioctl_flags & ZFS_NOUNLINK))
+		xoap->xoa_nounlink = B_TRUE;
+
+	if ((ioctl_flags & ZFS_REPARSE)) {
+		XVA_SET_REQ(xva, XAT_REPARSE);
+		xoap->xoa_reparse = B_TRUE;
+	}
```

Comment:
This code is effectively set-only.  It may be OK for now, since XAT_REPARSE setting seem to return EPERM, but for general case I guess it may be broken.

Reply:
Thank you for pointing it  out @amotin. I am working on it.

Reply:
@amotin updated to work for both set and unset.